google/cloud/spanner/integration_tests/instance_admin_integration_test.cc

```diff
@@ -39,88 +39,100 @@ using ::testing::AnyOf;
 using ::testing::HasSubstr;
 using ::testing::UnorderedElementsAre;
 
+std::string const& ProjectId() {
+  static std::string project_id =
+      internal::GetEnv("GOOGLE_CLOUD_PROJECT").value_or("");
+  return project_id;
+}
+
+std::string const& InstanceId() {
+  static std::string instance_id =
+      internal::GetEnv("GOOGLE_CLOUD_CPP_SPANNER_TEST_INSTANCE_ID")
+          .value_or("");
+  return instance_id;
+}
+
+bool RunSlowInstanceTests() {
+  static bool run_slow_instance_tests =
+      internal::GetEnv("GOOGLE_CLOUD_CPP_SPANNER_SLOW_INTEGRATION_TESTS")
+          .value_or("")
+          .find("instance") != std::string::npos;
+  return run_slow_instance_tests;
+}
+
+bool Emulator() {
+  static bool emulator = internal::GetEnv("SPANNER_EMULATOR_HOST").has_value();
+  return emulator;
+}
+
+class CleanupStaleInstances : public ::testing::Environment {
+ public:
+  void SetUp() override {
+    std::regex instance_name_regex(
+        R"(projects/.+/instances/)"
+        R"((temporary-instance-(\d{4}-\d{2}-\d{2})-.+))");
+
+    // Make sure we're using a correct regex.
+    EXPECT_EQ(2, instance_name_regex.mark_count());
+    auto generator = internal::MakeDefaultPRNG();
+    Instance in(ProjectId(), spanner_testing::RandomInstanceName(generator));
+    auto fq_instance_name = in.FullName();
+    std::smatch m;
+    EXPECT_TRUE(std::regex_match(fq_instance_name, m, instance_name_regex));
+    EXPECT_EQ(3, m.size());
+
+    EXPECT_STATUS_OK(spanner_testing::CleanupStaleInstances(
+        in.project_id(), instance_name_regex));
+  }
+};
+
+::testing::Environment* const kCleanupEnv =
+    ::testing::AddGlobalTestEnvironment(new CleanupStaleInstances);
+
 class InstanceAdminClientTest : public testing::Test {
  public:
-  InstanceAdminClientTest() : client_(MakeInstanceAdminConnection()) {}
+  InstanceAdminClientTest()
+      : generator_(internal::MakeDefaultPRNG()),
+        client_(MakeInstanceAdminConnection()) {
+    static_cast<void>(kCleanupEnv);
```

Comment:
So arguably we could do these with:

```cc
  static int const kCleanup = [] {
    // cleanup code here ...
  }();
```

Reply:
Arguably, yes (modulo any difference between running something when the first test is being constructed -v- before any test is constructed).

It looks like we have one other use of `testing::Environment`, which is doing what we handle here with the `static env_var` things here.  I'd be happy to try to eliminate both `testing::Environment` uses separately if you like.